src/main/java/com/aws/greengrass/tes/HttpServerImpl.java

```diff
@@ -5,51 +5,202 @@
+     * <p>The actual HTTP server port may be different from this,
+     * if configured port is system-assigned (0).
+     */
+    private final int configuredPort;
 
-    @Inject
+    private HttpServer serverIPv4;
+    private HttpServer serverIPv6;
+
+    private final HttpHandler credentialRequestHandler;
     private final ExecutorService executorService;
+    private final SocketFactory socketFactory;
+    private final HttpServerProvider httpServerProvider;
 
     /**
      * Constructor.
-     * @param port Http server port
+     * @param port HTTP server port
      * @param credentialRequestHandler request handler for server requests
      * @param executorService executor service instance
      * @throws IOException When server creation fails
      */
     HttpServerImpl(int port, HttpHandler credentialRequestHandler, ExecutorService executorService) throws IOException {
-        httpImpl = HttpServer.create(new InetSocketAddress("localhost", port), 0);
+        this(port, credentialRequestHandler, executorService,
+                HttpServerProvider.provider(), SocketFactory.getDefault());
+    }
+
+    HttpServerImpl(int port, HttpHandler credentialRequestHandler,
+                   ExecutorService executorService, HttpServerProvider httpServerProvider, SocketFactory socketFactory)
+            throws IOException {
         this.credentialRequestHandler = credentialRequestHandler;
         this.executorService = executorService;
+        this.configuredPort = port;
+        this.httpServerProvider = httpServerProvider;
+        this.socketFactory = socketFactory;
+        createServers(configuredPort);
+    }
+
+    private void createServers(int port) throws IOException {
+        // if port is system-picked, retry as we expect
+        // this to eventually succeed.
+        // otherwise, port is customer-configured,
+        // ensure we raise bind errors immediately.
+        if (port == 0) {
+            createServersWithRetry(port);
+        } else {
+            doCreateServers(port);
+        }
+    }
+
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.AvoidRethrowingException"})
+    private void createServersWithRetry(int port) throws IOException {
+        RetryUtils.RetryConfig retryConfig = RetryUtils.RetryConfig.builder()
+                .maxAttempt(Integer.MAX_VALUE)
+                .initialRetryInterval(Duration.ofSeconds(1L))
+                .maxRetryInterval(Duration.ofMinutes(1L))
+                .retryableExceptions(Collections.singletonList(SocketException.class))
+                .build();
+
+        try {
+            RetryUtils.runWithRetry(retryConfig,
+                    () -> {
+                        doCreateServers(port);
+                        return null;
+                    },
+                    "create-http-servers",
+                    logger);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)